### PR TITLE
Throwing NoContentException when InputStream is empty

### PR DESCRIPTION
**Pull request by mkarg**
Since JAX-RS 2.1 an MBR must react upon an empty stream either with an empty instance or with NoContentException, but Jersey's JSON-B MBR does neither - it simply throws ProcessingException:

_"In case the entity input stream is empty, the reader is expected to either return a Java representation of a zero-length entity or throw a javax.ws.rs.core.NoContentException in case no zero-length entity representation is defined for the supported Java type."_

### CODE REVIEW

**Comment by mkarg** [comment]
This PR proposes a solution which reconstitutes compliance with JAX-RS 2.1 with a two-step approach to detect an empty input stream:
* First it tries to use mark/reset as this is very fast, but not guaranteed to be supported by all kinds of input streams.
* If this is not possible, then it will actively buffer the first byte, which is rather slow, but will work with all kinds of input stream.

As long as we allow this MBR to be used with other JAX-RS implementations (not just Jersey), we need this PR, and we need this two-step approach. An alternative solution could be to clearly restrict this MBR to work on Jersey only (e. g. by documentation or even technical checks), in which case we could leave out one of the two steps as we then know for sure the type of input stream Jersey will run this MBR with.

**Comment by mkarg** [comment]
Please see also https://github.com/eclipse-ee4j/jaxrs-api/issues/795.

**Comment by mkarg** [comment]
@spericas @chkal I assume this is the code location where you want the topic discussed in https://github.com/eclipse-ee4j/jaxrs-api/issues/795 to be handled (rather than the JAX-RS spec). :-)

**Comment by jansupol** [comment]
Is it possible to add a test checking the behaviour for an empty stream?

File: media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java [hunk 1]
```diff
@@ -70,7 +72,24 @@ public Object readFrom(Class<Object> type, Type genericType,
...
+            entityStream.mark(1);
+            if (entityStream.read() == -1) {
+                throw new NoContentException("JSON-B cannot process empty input stream");
+            }
+            entityStream.reset();
+        } else {
+            final PushbackInputStream buffer = new PushbackInputStream(entityStream);
+            final int firstByte = buffer.read();
+            if (firstByte == -1) {
+                throw new NoContentException("JSON-B cannot process empty input stream");
```
Comment (jansupol) [inline media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java]: Please use Localization.properties

File: media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java [hunk 2]
```diff
@@ -70,7 +72,24 @@ public Object readFrom(Class<Object> type, Type genericType,
                            MediaType mediaType,
                            MultivaluedMap<String, String> httpHeaders,
                            InputStream entityStream) throws IOException, WebApplicationException {
+        if (entityStream.markSupported()) {
+            entityStream.mark(1);
+            if (entityStream.read() == -1) {
+                throw new NoContentException("JSON-B cannot process empty input stream");
```
Comment (jansupol) [inline media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java]: Please use localization.properties

**Comment by mkarg** [comment]
> Is it possible to add a test checking the behaviour for an empty stream?

Added unit test covering this issue with 4c6b2acadac2190daa783295eaa9ef2fc694d57e.

File: media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java (hunk 1 above)
Comment (mkarg) [reply media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java]: I will do that. Stay tuned.

File: media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java (hunk 2 above)
Comment (mkarg) [reply media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java]: I will do that. Stay tuned.

File: media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java (hunk 1 above)
Comment (mkarg) [reply media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java]: Done with bccd9008528477587b2eb340cbaf0a8762bcdadb.

File: media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java (hunk 2 above)
Comment (mkarg) [reply media/json-binding/src/main/java/org/glassfish/jersey/jsonb/internal/JsonBindingProvider.java]: Done with bccd9008528477587b2eb340cbaf0a8762bcdadb.

**Comment by senivam** [comment]
just another thought... why not to align everything to a Stream which supports marks and to separate to a private utility method? So, the main method would look like:
```
@Override
    public Object readFrom(Class<Object> type, Type genericType,
                           Annotation[] annotations,
                           MediaType mediaType,
                           MultivaluedMap<String, String> httpHeaders,
                           InputStream entityStream) throws IOException, WebApplicationException {

        validateInputStream(entityStream);

        Jsonb jsonb = getJsonb(type);

        try {
            return jsonb.fromJson(entityStream, genericType);
        } catch (JsonbException e) {
            throw new ProcessingException(LocalizationMessages.ERROR_JSONB_DESERIALIZATION(), e);
        }
    }
```
and then private utility method:
```
    private static void validateInputStream(InputStream entityStream) throws IOException {
        final InputStream streamWithMark = (entityStream.markSupported()) ? entityStream : new BufferedInputStream(entityStream);

        streamWithMark.mark(1);
        if (streamWithMark.read() == -1) {
            throw new NoContentException(LocalizationMessages.ERROR_JSONB_EMPTYSTREAM());
        }
        streamWithMark.reset();
        entityStream = streamWithMark;
    }
```
what do you think?

**Comment by mkarg** [comment]
> just another thought... _previous comment was wrong :)_

Cool optimization! 😎 If you are sure it is as safe and as fast I would be happy to accept your PR ontop! 👍

**Comment by senivam** [comment]
@mkarg, OK, I've submitted a PR to your branch :)
